diff --git a/src/etc/inc/interfaces.inc b/src/etc/inc/interfaces.inc
index 11cbd9d..345f166 100644
--- a/src/etc/inc/interfaces.inc
+++ b/src/etc/inc/interfaces.inc
@@ -1184,6 +1184,11 @@ function interface_is_type($if, $type) {
 			$entry = 'lagg';
 			$entif = 'laggif';
 			break;
+		case "vxlan":
+			$list = 'vxlans';
+			$entry = 'vxlan';
+			$entif = 'vxlanif';
+			break;
 		default:
 			break;
 	}
@@ -1361,8 +1366,73 @@ function interface_gif_configure(&$gif, $gifkey = "", $flush = true) {
 	return $gifif;
 }
 
+/* NOTE: $vxlankey is not used but useful for passing this function to array_walk. */
+function interface_vxlan_configure(&$vxlan, $vxlankey = "", $flush = true) {
+	global $config, $g;
+	if (!is_array($vxlan)) {
+		return -1;
+	}
+	$realif = convert_friendly_interface_to_real_interface_name($vxlan['if']);
+	if (!interface_is_vlan($realif)) {
+		$realif = get_real_interface($vxlan['if']);
+	}
+	$realifip = get_interface_ip($vxlan['if']);
+	$realifip6 = get_interface_ipv6($vxlan['if']);
+	/* make sure the parent interface is up */
+	if ($realif) {
+		interfaces_bring_up($realif);
+	} else {
+		log_error(gettext("could not bring parentif up -- variable not defined -- interface_vxlan_configure()"));
+	}
+	/* Interface is defined */
+	if (platform_booting() || !(empty($vxlan['vxlanif']))) {
+		pfSense_interface_destroy($vxlan['vxlanif']);
+		pfSense_interface_create2($vxlan['vxlanif']);
+		$vxlanif = $vxlan['vxlanif'];
+	} else {
+		$vxlanif = pfSense_interface_create2("vxlan");
+	}
+	$ifconfigcmd = "/sbin/ifconfig " . escapeshellarg($vxlanif) .
+	    " vxlanid " .  escapeshellarg($vxlan['vxlanid']);
+	if (is_mcast($vxlan['remote-addr'])) {
+		$ifconfigcmd .= " vxlangroup " .
+		    escapeshellarg($vxlan['remote-addr']) . " vxlandev " .
+		    escapeshellarg($realif);
+	} else {
+		$ifconfigcmd .= " vxlanremote " .
+		    escapeshellarg($vxlan['remote-addr']);
+	}
+	if (is_ipaddrv6($vxlan['remote-addr']) && $realifip6) {
+		$ifconfigcmd .= " vxlanlocal " . escapeshellarg($realifip6);
+	} elseif (is_ipaddrv4($vxlan['remote-addr']) && $realifip) {
+		$ifconfigcmd .= " vxlanlocal " . escapeshellarg($realifip);
+	} else {
+		return;
+	}
+	if (!empty($vxlan['vxlanlocalport'])) {
+		$ifconfigcmd .= " vxlanlocalport " . escapeshellarg($vxlan['vxlanlocalport']);
+	}
+	if (!empty($vxlan['vxlanremoteport'])) {
+		$ifconfigcmd .= " vxlanremoteport " . escapeshellarg($vxlan['vxlanremoteport']);
+	}
+	if (!empty($vxlan['vxlanttl'])) {
+		$ifconfigcmd .= " vxlanttl " . escapeshellarg($vxlan['vxlanttl']);
+	}
+	if (isset($vxlan['vxlanlearn'])) {
+		$ifconfigcmd .= " vxlanlearn ";
+	} else {
+		$ifconfigcmd .= " -vxlanlearn ";
+	}
+	mwexec($ifconfigcmd);
+	if ($flush) {
+		get_interface_arr(true);
+	}
+	interfaces_bring_up($vxlanif);
+	return $vxlanif;
+}
+
 function interfaces_tunnel_configure($checkparent = 0, $realif = "", $type = "") {
-	if (!in_array($type, array('gre', 'gif'))) {
+	if (!in_array($type, array('gre', 'gif', 'vxlan'))) {
 		return;
 	}
 
@@ -1402,6 +1472,8 @@ function interfaces_tunnel_configure($checkparent = 0, $realif = "", $type = "")
 			interface_gif_configure($tunnel, "", false);
 		} elseif ($type == 'gre') {
 			interface_gre_configure($tunnel, "", false);
+		} elseif ($type == 'vxlan') {
+			interface_vxlan_configure($tunnel, "", false);
 		}
 	}
 
@@ -1652,6 +1724,8 @@ function interfaces_configure() {
 			$delayed_list[$if] = $ifname;
 		} elseif (strstr($realif, "gif")) {
 			$delayed_list[$if] = $ifname;
+		} elseif (strstr($realif, "vxlan")) {
+			$delayed_list[$if] = $ifname;
 		} elseif (strstr($realif, "ovpn")) {
 			continue;
 		} elseif (strstr($realif, "ipsec")) {
@@ -1697,6 +1771,9 @@ function interfaces_configure() {
 	/* set up GIF virtual interfaces */
 	interfaces_tunnel_configure(1,'','gif');
 
+	/* set up VXLAN virtual interfaces */
+	interfaces_tunnel_configure(1,'','vxlan');
+
 	/* set up BRIDGE virtual interfaces */
 	interfaces_bridge_configure(1);
 
@@ -1724,6 +1801,9 @@ function interfaces_configure() {
 	/* set up GIF virtual interfaces */
 	interfaces_tunnel_configure(2,'','gif');
 
+	/* set up VXLAN virtual interfaces */
+	interfaces_tunnel_configure(2,'','vxlan');
+
 	foreach ($delayed_list as $if => $ifname) {
 		if (is_platform_booting()) {
 			printf(gettext("Configuring %s interface..."), $ifname);
@@ -4323,6 +4403,10 @@ function interface_configure($interface = "wan", $reloadall = false, $linkupeven
 		$wantedmtu = 1280;
 	}
 
+	if (($wantedmtu == 0) && interface_is_type($mtuif, 'vxlan')) {
+		$wantedmtu = 1450;
+	}
+
 	/* Set the MTU to 1500 if no explicit MTU configured. */
 	if ($wantedmtu == 0) {
 		$wantedmtu = 1500; /* Default */
@@ -4387,7 +4471,7 @@ function interface_configure($interface = "wan", $reloadall = false, $linkupeven
 			break;
 		default:
 			/* XXX: Kludge for now related to #3280 */
-			if (!in_array($tunnelif, array("gif", "gre", "ovp", "ips", "l2t"))) {
+			if (!in_array($tunnelif, array("gif", "gre", "vxlan", "ovp", "ips", "l2t"))) {
 				if (is_ipaddrv4($wancfg['ipaddr']) && $wancfg['subnet'] <> "") {
 					pfSense_interface_setaddress($realif, "{$wancfg['ipaddr']}/{$wancfg['subnet']}");
 				}
@@ -4419,7 +4503,7 @@ function interface_configure($interface = "wan", $reloadall = false, $linkupeven
 			break;
 		default:
 			/* XXX: Kludge for now related to #3280 */
-			if (!in_array($tunnelif, array("gif", "gre", "ovp", "ips", "l2t"))) {
+			if (!in_array($tunnelif, array("gif", "gre", "vxlan", "ovp", "ips", "l2t"))) {
 				if (is_ipaddrv6($wancfg['ipaddrv6']) && $wancfg['subnetv6'] <> "") {
 					//pfSense_interface_setaddress($realif, "{$wancfg['ipaddrv6']}/{$wancfg['subnetv6']}");
 					// FIXME: Add IPv6 Support to the pfSense module
@@ -4449,6 +4533,11 @@ function interface_configure($interface = "wan", $reloadall = false, $linkupeven
 			array_walk($gif, 'interface_gif_configure');
 		}
 
+		if ($tunnelif != 'vxlan') {
+			$vxlan = link_interface_to_tunnelif($interface, 'vxlan');
+			array_walk($vxlan, 'interface_vxlan_configure');
+		}
+
 		$grouptmp = link_interface_to_group($interface);
 		if (!empty($grouptmp)) {
 			array_walk($grouptmp, 'interface_group_add_member');
@@ -6406,7 +6495,7 @@ function link_interface_to_tunnelif($interface, $type, $remote = 'any') {
 		return $result;
 	}
 
-	if (!in_array($type, array('gre', 'gif'))) {
+	if (!in_array($type, array('gre', 'gif', 'vxlan'))) {
 		return $result;
 	}
 
@@ -7578,7 +7667,7 @@ function create_interface_list($open = false) {
 }
 
 function is_pseudo_interface($inf, $tap=true) {
-	$psifs = array('ovpn', 'ipsec', 'l2tp', 'pptp', 'gif', 'gre', 'ppp', 'pppoe');
+	$psifs = array('ovpn', 'ipsec', 'l2tp', 'pptp', 'gif', 'gre', 'ppp', 'pppoe', 'vxlan');
 	foreach ($psifs as $pif) {
 		if (substr($inf, 0, strlen($pif)) == $pif) {
 			if (($pif == 'ovpn') && $tap) {
diff --git a/src/etc/inc/interfaces_fast.inc b/src/etc/inc/interfaces_fast.inc
index a102fe8..e35041f 100644
--- a/src/etc/inc/interfaces_fast.inc
+++ b/src/etc/inc/interfaces_fast.inc
@@ -246,6 +246,11 @@ function interface_assign_description_fast($portlist, $friendlyifnames) {
 			if ($portinfo['descr']) {
 				$descr .= " (" . $portinfo['descr'] . ")";
 			}
+		} elseif ($portinfo['isvxlan']) {
+			$descr = "VXLAN {$portinfo['remote-addr']}";
+			if ($portinfo['descr']) {
+				$descr .= " (" . $portinfo['descr'] . ")";
+			}
 		} elseif ($portinfo['islagg']) {
 			$descr = strtoupper($portinfo['laggif']);
 			if ($portinfo['descr']) {
diff --git a/src/etc/inc/util.inc b/src/etc/inc/util.inc
index 3c395b1..7319d16 100644
--- a/src/etc/inc/util.inc
+++ b/src/etc/inc/util.inc
@@ -1985,6 +1985,7 @@ function get_interface_list($mode = "active", $keyby = "physical", $vfaces = fal
 				'plip',
 				'pfsync',
 				'enc',
+				'vxlan',
 				'tun',
 				'lagg',
 				'vip',
@@ -2951,7 +2952,7 @@ function is_interface_mismatch() {
 		foreach (config_get_path('interfaces', []) as $ifcfg) {
 			if (interface_is_vlan($ifcfg['if']) != NULL ||
 			    interface_is_qinq($ifcfg['if']) != NULL ||
-			    preg_match("/^bridge|^cua|^enc|^gif|^gre|^ipsec|^l2tp|^lagg|^ngeth|^ovpn|^ppp|^pptp|^tap|^tun|^ue|vlan|^wg|_wlan|_\d{0,4}_\d{0,4}$/i", $ifcfg['if'])) {
+			    preg_match("/^bridge|^cua|^enc|^gif|^gre|^ipsec|^l2tp|^lagg|^ngeth|^ovpn|^ppp|^pptp|^tap|^tun|^ue|vlan|^wg|^vxlan|_wlan|_\d{0,4}_\d{0,4}$/i", $ifcfg['if'])) {
 				// Do not check these interfaces.
 				$i++;
 				continue;
diff --git a/src/etc/rc.newwanip b/src/etc/rc.newwanip
index 46b6396..58dad06 100755
--- a/src/etc/rc.newwanip
+++ b/src/etc/rc.newwanip
@@ -132,6 +132,9 @@ array_walk($gre, 'interface_gre_configure');
 $gif = link_interface_to_tunnelif($interface, 'gif', 'inet');
 array_walk($gif, 'interface_gif_configure');
 
+$vxlan = link_interface_to_tunnelif($interface, 'vxlan', 'inet');
+array_walk($vxlan, 'interface_vxlan_configure');
+
 /* reconfigure L2TP/PPTP tunnels, see https://redmine.pfsense.org/issues/12072 */ 
 $ppp = link_interface_to_ppp_tunnelif($interface);
 
@@ -188,6 +191,13 @@ foreach ($gre as $gretun) {
 		system_routing_configure($confif);
 	}
 }
+foreach ($vxlan as $vxlantun) {
+	$confif = convert_real_interface_to_friendly_interface_name($vxlantun['vxlanif']);
+	if (!empty($confif)) {
+		interface_configure($confif);
+		system_routing_configure($confif);
+	}
+}
 foreach ($ppp as $ppptun) {
 	$confif = convert_real_interface_to_friendly_interface_name($ppptun['if']);
 	if (!empty($confif)) {
diff --git a/src/usr/local/www/interfaces_assign.php b/src/usr/local/www/interfaces_assign.php
index e33a1bc..e782cc8 100644
--- a/src/usr/local/www/interfaces_assign.php
+++ b/src/usr/local/www/interfaces_assign.php
@@ -98,6 +98,12 @@ foreach (config_get_path('gres/gre', []) as $gre) {
 	$portlist[$gre['greif']]['isgre'] = true;
 }
 
+/* add VXLAN interfaces */
+foreach (config_get_path('vxlans/vxlan', []) as $vxlan) {
+	$portlist[$vxlan['vxlanif']] = $vxlan;
+	$portlist[$vxlan['vxlanif']]['isvxlan'] = true;
+}
+
 /* add LAGG interfaces */
 foreach (config_get_path('laggs/lagg', []) as $lagg) {
 	$portlist[$lagg['laggif']] = $lagg;
@@ -299,7 +305,8 @@ if (isset($_REQUEST['add']) && isset($_REQUEST['if_add'])) {
 					}
 
 					if ((substr($ifport, 0, 3) == 'gre') ||
-					    (substr($ifport, 0, 5) == 'gif')) {
+					    (substr($ifport, 0, 5) == 'gif') ||
+						(substr($ifport, 0, 5) == 'vxlan')) {
 						unset($this_if_config['ipaddr']);
 						unset($this_if_config['subnet']);
 						unset($this_if_config['ipaddrv6']);
@@ -362,6 +369,8 @@ if (isset($_REQUEST['add']) && isset($_REQUEST['if_add'])) {
 			$input_errors[] = gettext("The interface is part of a gre tunnel. Please delete the tunnel to continue");
 		} else if (!empty(link_interface_to_tunnelif($id, 'gif'))) {
 			$input_errors[] = gettext("The interface is part of a gif tunnel. Please delete the tunnel to continue");
+		} else if (!empty(link_interface_to_tunnelif($id, 'vxlan'))) {
+			$input_errors[] = gettext("The interface is part of a vxlan tunnel. Please delete the tunnel to continue");
 		} else if (interface_has_queue($id)) {
 			$input_errors[] = gettext("The interface has a traffic shaper queue configured.\nPlease remove all queues on the interface to continue.");
 		} else {
@@ -467,6 +476,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_bridge.php b/src/usr/local/www/interfaces_bridge.php
index 16434ab..73889d1 100644
--- a/src/usr/local/www/interfaces_bridge.php
+++ b/src/usr/local/www/interfaces_bridge.php
@@ -83,6 +83,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), true, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_gif.php b/src/usr/local/www/interfaces_gif.php
index 6011305..c6742e5 100644
--- a/src/usr/local/www/interfaces_gif.php
+++ b/src/usr/local/www/interfaces_gif.php
@@ -79,6 +79,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), true, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_gre.php b/src/usr/local/www/interfaces_gre.php
index 557b777..ab0e4bc 100644
--- a/src/usr/local/www/interfaces_gre.php
+++ b/src/usr/local/www/interfaces_gre.php
@@ -79,6 +79,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), true, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_groups.php b/src/usr/local/www/interfaces_groups.php
index 037c28c..4dbdc7e 100644
--- a/src/usr/local/www/interfaces_groups.php
+++ b/src/usr/local/www/interfaces_groups.php
@@ -61,6 +61,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_lagg.php b/src/usr/local/www/interfaces_lagg.php
index 7b24452..3e12307 100644
--- a/src/usr/local/www/interfaces_lagg.php
+++ b/src/usr/local/www/interfaces_lagg.php
@@ -84,6 +84,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), true, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_ppps.php b/src/usr/local/www/interfaces_ppps.php
index 814141b..23f7cc4 100644
--- a/src/usr/local/www/interfaces_ppps.php
+++ b/src/usr/local/www/interfaces_ppps.php
@@ -87,6 +87,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), true, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_qinq.php b/src/usr/local/www/interfaces_qinq.php
index 68e6c7f..687cc24 100644
--- a/src/usr/local/www/interfaces_qinq.php
+++ b/src/usr/local/www/interfaces_qinq.php
@@ -96,6 +96,7 @@ $tab_array[] = array(gettext("QinQs"), true, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_vlan.php b/src/usr/local/www/interfaces_vlan.php
index 17f9c07..e3dc00c 100644
--- a/src/usr/local/www/interfaces_vlan.php
+++ b/src/usr/local/www/interfaces_vlan.php
@@ -91,6 +91,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
diff --git a/src/usr/local/www/interfaces_vlan_edit.php b/src/usr/local/www/interfaces_vlan_edit.php
index bcde5d1..a54bd80 100644
--- a/src/usr/local/www/interfaces_vlan_edit.php
+++ b/src/usr/local/www/interfaces_vlan_edit.php
@@ -62,6 +62,17 @@ foreach ($portlist as $portname => $port) {
 	}
 }
 
+/* allow to select VXLAN interfaces, see https://redmine.pfsense.org/issues/11143 */
+if (is_array($config['vxlans']['vxlan']) && count($config['vxlans']['vxlan'])) {
+	foreach ($config['vxlans']['vxlan'] as $vxlan) {
+		$portlist[$vxlan['vxlanif']] = array('mac' => $vxlan['vxlanif']);
+		$friendly = convert_real_interface_to_friendly_interface_name($vxlan['vxlanif']);
+		if ($friendly) {
+			$portlist[$vxlan['vxlanif']]['friendly'] = $friendly;
+		}
+	}
+}
+
 if (is_numericint($_REQUEST['id'])) {
 	$id = $_REQUEST['id'];
 }
diff --git a/src/usr/local/www/interfaces_vxlan.php b/src/usr/local/www/interfaces_vxlan.php
new file mode 100644
index 0000000..426dde2
--- /dev/null
+++ b/src/usr/local/www/interfaces_vxlan.php
@@ -0,0 +1,140 @@
+<?php
+/*
+ * interfaces_vxlan.php
+ *
+ * part of pfSense (https://www.pfsense.org)
+ * Copyright (c) 2004-2013 BSD Perimeter
+ * Copyright (c) 2013-2016 Electric Sheep Fencing
+ * Copyright (c) 2014-2021 Rubicon Communications, LLC (Netgate)
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+##|+PRIV
+##|*IDENT=page-interfaces-vxlan
+##|*NAME=Interfaces: VXLAN
+##|*DESCR=Allow access to the 'Interfaces: VXLAN' page.
+##|*MATCH=interfaces_vxlan.php*
+##|-PRIV
+
+require_once("guiconfig.inc");
+require_once("functions.inc");
+
+init_config_arr(array('vxlans', 'vxlan'));
+$a_vxlans = &$config['vxlans']['vxlan'] ;
+
+function vxlan_inuse($num) {
+	global $config, $a_vxlans;
+
+	$iflist = get_configured_interface_list(true);
+	foreach ($iflist as $if) {
+		if ($config['interfaces'][$if]['if'] == $a_vxlans[$num]['vxlanif']) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+if ($_POST['act'] == "del") {
+	if (!isset($_POST['id'])) {
+		$input_errors[] = gettext("Wrong parameters supplied");
+	} else if (empty($a_vxlans[$_POST['id']])) {
+		$input_errors[] = gettext("Wrong index supplied");
+	/* check if still in use */
+	} else if (vxlan_inuse($_POST['id'])) {
+		$input_errors[] = gettext("This VXLAN tunnel cannot be deleted because it is still being used as an interface.");
+	} else {
+		pfSense_interface_destroy($a_vxlans[$_POST['id']]['vxlanif']);
+		unset($a_vxlans[$_POST['id']]);
+
+		write_config("VXLAN interface deleted");
+
+		header("Location: interfaces_vxlan.php");
+		exit;
+	}
+}
+
+$pgtitle = array(gettext("Interfaces"), gettext("VXLANs"));
+$shortcut_section = "interfaces";
+include("head.inc");
+if ($input_errors) {
+	print_input_errors($input_errors);
+}
+
+$tab_array = array();
+$tab_array[] = array(gettext("Interface Assignments"), false, "interfaces_assign.php");
+$tab_array[] = array(gettext("Interface Groups"), false, "interfaces_groups.php");
+$tab_array[] = array(gettext("Wireless"), false, "interfaces_wireless.php");
+$tab_array[] = array(gettext("VLANs"), false, "interfaces_vlan.php");
+$tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
+$tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
+$tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
+$tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), true, "interfaces_vxlan.php");
+$tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
+$tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
+display_top_tabs($tab_array);
+?>
+<div class="panel panel-default">
+	<div class="panel-heading"><h2 class="panel-title"><?=gettext('VXLAN Interfaces')?></h2></div>
+	<div class="panel-body">
+		<div class="table-responsive">
+			<table class="table table-striped table-hover table-condensed table-rowdblclickedit">
+				<thead>
+					<tr>
+						<th><?=gettext("Interface"); ?></th>
+						<th><?=gettext("Tunnel to"); ?></th>
+						<th><?=gettext("Description"); ?></th>
+						<th><?=gettext("Actions"); ?></th>
+					</tr>
+				</thead>
+				<tbody>
+<?php foreach ($a_vxlans as $i => $vxlan):
+	if (substr($vxlan['if'], 0, 4) == "_vip") {
+		$if = convert_real_interface_to_friendly_descr(get_real_interface($vxlan['if']));
+	} else {
+		$if = $vxlan['if'];
+	}
+?>
+					<tr>
+						<td>
+							<?=htmlspecialchars(convert_friendly_interface_to_friendly_descr($if))?>
+						</td>
+						<td>
+							<?=htmlspecialchars($vxlan['remote-addr'])?>
+						</td>
+						<td>
+							<?=htmlspecialchars($vxlan['descr'])?>
+						</td>
+						<td>
+							<a class="fa fa-pencil"	title="<?=gettext('Edit VXLAN interface')?>"	href="interfaces_vxlan_edit.php?id=<?=$i?>"></a>
+							<a class="fa fa-trash"	title="<?=gettext('Delete VXLAN interface')?>"	href="interfaces_vxlan.php?act=del&amp;id=<?=$i?>" usepost></a>
+						</td>
+					</tr>
+<?php endforeach; ?>
+				</tbody>
+			</table>
+		</div>
+	</div>
+</div>
+
+<nav class="action-buttons">
+	<a href="interfaces_vxlan_edit.php" class="btn btn-success btn-sm">
+		<i class="fa fa-plus icon-embed-btn"></i>
+		<?=gettext("Add")?>
+	</a>
+</nav>
+<?php
+include("foot.inc");
diff --git a/src/usr/local/www/interfaces_vxlan_edit.php b/src/usr/local/www/interfaces_vxlan_edit.php
new file mode 100644
index 0000000..ea0cf23
--- /dev/null
+++ b/src/usr/local/www/interfaces_vxlan_edit.php
@@ -0,0 +1,248 @@
+<?php
+/*
+ * interfaces_vxlan_edit.php
+ *
+ * part of pfSense (https://www.pfsense.org)
+ * Copyright (c) 2020-2021 Rubicon Communications, LLC (Netgate)
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+##|+PRIV
+##|*IDENT=page-interfaces-vxlan-edit
+##|*NAME=Interfaces: VXLAN: Edit
+##|*DESCR=Allow access to the 'Interfaces: VXLAN: Edit' page.
+##|*MATCH=interfaces_vxlan_edit.php*
+##|-PRIV
+
+require_once("guiconfig.inc");
+
+init_config_arr(array('vxlans', 'vxlan'));
+$a_vxlans = &$config['vxlans']['vxlan'];
+$id = $_REQUEST['id'];
+
+
+if (isset($id) && $a_vxlans[$id]) {
+	$pconfig['if'] = $a_vxlans[$id]['if'];
+	if (!empty($a_vxlans[$id]['ipaddr'])) {
+		$pconfig['if'] = $pconfig['if'] . '|' . $a_vxlans[$id]['ipaddr'];
+	}
+	$pconfig['vxlanif'] = $a_vxlans[$id]['vxlanif'];
+	$pconfig['remote-addr'] = $a_vxlans[$id]['remote-addr'];
+	$pconfig['vxlanid'] = $a_vxlans[$id]['vxlanid'];
+	$pconfig['vxlanlocalport'] = $a_vxlans[$id]['vxlanlocalport'];
+	$pconfig['vxlanremoteport'] = $a_vxlans[$id]['vxlanremoteport'];
+	$pconfig['vxlanttl'] = $a_vxlans[$id]['vxlanttl'];
+	$pconfig['vxlanlearn'] = isset($a_vxlans[$id]['vxlanlearn']);
+	$pconfig['descr'] = $a_vxlans[$id]['descr'];
+} else {
+	$pconfig['vxlanlearn'] = true;
+}
+
+if ($_POST['save']) {
+	unset($input_errors);
+	$pconfig = $_POST;
+
+	/* input validation */
+	$reqdfields = explode(" ", "if remote-addr vxlanid");
+	$reqdfieldsn = array(gettext("Parent interface"), gettext("VXLAN remote address"), gettext("VXLAN ID"));
+
+	do_input_validation($_POST, $reqdfields, $reqdfieldsn, $input_errors);
+
+	if (!is_ipaddr($_POST['remote-addr']) || is_subnet($_POST['remote-addr'])) {
+		$input_errors[] = gettext("The remote peer address must have valid IP address and must not contain CIDR masks or prefixes.");
+	}
+
+	$alias = strstr($_POST['if'], '|');
+	if ((is_ipaddrv4($alias) && !is_ipaddrv4($_POST['remote-addr'])) ||
+	    (is_ipaddrv6($alias) && !is_ipaddrv6($_POST['remote-addr']))) {
+		$input_errors[] = gettext("The alias IP address family has to match the family of the remote peer address.");
+	}
+
+	if (!is_numeric($_POST['vxlanid']) || ($_POST['vxlanid'] < 1) || ($_POST['vxlanid'] > 16777215)) {
+		$input_errors[] = gettext("VXLAN ID valid range is 1-16777215.");
+	}
+
+	if (!empty($_POST['vxlanlocalport']) && !is_port($_POST['vxlanlocalport'])) {
+		$input_errors[] = gettext("VXLAN Local Port must be valid port value.");
+	}
+
+	if (!empty($_POST['vxlanremoteport']) && !is_port($_POST['vxlanremoteport'])) {
+		$input_errors[] = gettext("VXLAN Local Port must be valid port value.");
+	}
+
+	if (!empty($_POST['vxlanttl']) && (!is_numeric($_POST['vxlanttl']) ||
+	    ($_POST['vxlanttl'] < 1) || ($_POST['vxlanttl'] > 255))) {
+		$input_errors[] = gettext("VXLAN TTL must be in range between 1 and 255.");
+	}
+
+	foreach ($a_vxlans as $vxlan) {
+		if (isset($id) && ($a_vxlans[$id]) && ($a_vxlans[$id] === $vxlan)) {
+			continue;
+		}
+
+		if (($vxlan['if'] == $_POST['if']) && ($vxlan['remote-addr'] == $_POST['remote-addr']) &&
+		    ($vxlan['vxlanid'] == $vxlan['vxlanid'])) {
+			$input_errors[] = sprintf(gettext("A vxlan with the remote address %s and VXLAN ID %s is already defined."), $vxlan['remote-addr'], $vxlan['vxlanid']);
+			break;
+		}
+	}
+
+	if (!$input_errors) {
+		$vxlan = array();
+		list($vxlan['if'], $vxlan['ipaddr']) = explode("|", $_POST['if']);
+		$vxlan['remote-addr'] = $_POST['remote-addr'];
+		$vxlan['vxlanid'] = $_POST['vxlanid'];
+		$vxlan['vxlanlocalport'] = $_POST['vxlanlocalport'];
+		$vxlan['vxlanremoteport'] = $_POST['vxlanremoteport'];
+		$vxlan['vxlanttl'] = $_POST['vxlanttl'];
+		if (isset($_POST['vxlanlearn'])) {
+			$vxlan['vxlanlearn'] = '';
+		}
+		$vxlan['descr'] = $_POST['descr'];
+		$vxlan['vxlanif'] = $_POST['vxlanif'];
+		$vxlan['vxlanif'] = interface_vxlan_configure($vxlan);
+
+		if ($vxlan['vxlanif'] == "" || !stristr($vxlan['vxlanif'], "vxlan")) {
+			$input_errors[] = gettext("Error occurred creating interface, please retry.");
+		} else {
+			if (isset($id) && $a_vxlans[$id]) {
+				$a_vxlans[$id] = $vxlan;
+			} else {
+				$a_vxlans[] = $vxlan;
+			}
+
+			write_config("VXLAN interface added");
+
+			$confif = convert_real_interface_to_friendly_interface_name($vxlan['vxlanif']);
+
+			if ($confif != "") {
+				interface_configure($confif);
+			}
+
+			header("Location: interfaces_vxlan.php");
+			exit;
+		}
+	}
+}
+
+function build_parent_list() {
+	$parentlist = array();
+	$portlist = get_possible_listen_ips();
+	foreach ($portlist as $ifn => $ifinfo) {
+		$parentlist[$ifn] = $ifinfo;
+	}
+
+	return($parentlist);
+}
+
+$pgtitle = array(gettext("Interfaces"), gettext("VXLANs"), gettext("Edit"));
+$pglinks = array("", "interfaces_vxlan.php", "@self");
+$shortcut_section = "interfaces";
+include("head.inc");
+
+if ($input_errors) {
+	print_input_errors($input_errors);
+}
+
+$form = new Form();
+
+$section = new Form_Section('VXLAN Configuration');
+
+$section->addInput(new Form_Select(
+	'if',
+	'*Parent Interface',
+	$pconfig['if'],
+	build_parent_list()
+))->setHelp('This interface serves as the local address to be used for the VXLAN tunnel.');
+
+$section->addInput(new Form_IpAddress(
+	'remote-addr',
+	'*VXLAN Remote Address',
+	$pconfig['remote-addr']
+))->setHelp('Peer address or multicast group where encapsulated VXLAN packets will be sent.');
+
+$section->addInput(new Form_Input(
+	'vxlanid',
+	'*VXLAN ID',
+	'text',
+	$pconfig['vxlanid']
+))->setHelp('24-bit VXLAN Network Identifier (VNI) that identifies the virtual network segment membership of the interface.');
+
+$section->addInput(new Form_Input(
+	'vxlanlocalport',
+	'VXLAN Local Port',
+	'text',
+	$pconfig['vxlanlocalport'],
+	['placeholder' => '4789']
+))->setHelp('The port number the interface will listen on. The default port number is 4789.');
+
+$section->addInput(new Form_Input(
+	'vxlanremoteport',
+	'VXLAN Remote Port',
+	'text',
+	$pconfig['vxlanremoteport'],
+	['placeholder' => '4789']
+))->setHelp('The destination port number used in the encapsulating IPv4/IPv6 header. ' .
+	'The remote host should be listening on this port. The default port number is 4789. ' .
+	'Note some other implementations, such as Linux, do not default to the IANA assigned port, ' .
+	'but instead listen on port 8472.');
+
+$section->addInput(new Form_Input(
+	'vxlanttl',
+	'VXLAN TTL',
+	'text',
+	$pconfig['vxlanttl'],
+	['placeholder' => '64']
+))->setHelp('The TTL used in the encapsulating IPv4/IPv6 header. The default is	64.');
+
+$section->addInput(new Form_Checkbox(
+	'vxlanlearn',
+	'VXLAN Learn',
+	'Populate the forwarding table with the source IP address and inner source Ethernet MAC address of received packets.',
+	$pconfig['vxlanlearn']
+))->setHelp('The source IP address and inner source Ethernet MAC address of ' .
+	'received packets are used to dynamically populate the forwarding ' .
+	'table. When in multicast mode, an entry in the forwarding table ' .
+	'allows the interface to send the frame directly to the remote ' .
+	'host instead of broadcasting the frame to the multicast group. ' .
+	'This is the default.');
+
+$section->addInput(new Form_Input(
+	'descr',
+	'Description',
+	'text',
+	$pconfig['descr']
+))->setHelp('A description may be entered here for administrative reference (not parsed).');
+
+$form->addGlobal(new Form_Input(
+	'vxlanif',
+	null,
+	'hidden',
+	$pconfig['vxlanif']
+));
+
+if (isset($id) && $a_vxlans[$id]) {
+	$form->addGlobal(new Form_Input(
+		'id',
+		null,
+		'hidden',
+		$id
+	));
+}
+
+$form->add($section);
+print($form);
+
+include("foot.inc");
diff --git a/src/usr/local/www/interfaces_wireless.php b/src/usr/local/www/interfaces_wireless.php
index 6bde9d4..e349afe 100644
--- a/src/usr/local/www/interfaces_wireless.php
+++ b/src/usr/local/www/interfaces_wireless.php
@@ -77,6 +77,7 @@ $tab_array[] = array(gettext("QinQs"), false, "interfaces_qinq.php");
 $tab_array[] = array(gettext("PPPs"), false, "interfaces_ppps.php");
 $tab_array[] = array(gettext("GREs"), false, "interfaces_gre.php");
 $tab_array[] = array(gettext("GIFs"), false, "interfaces_gif.php");
+$tab_array[] = array(gettext("VXLANs"), false, "interfaces_vxlan.php");
 $tab_array[] = array(gettext("Bridges"), false, "interfaces_bridge.php");
 $tab_array[] = array(gettext("LAGGs"), false, "interfaces_lagg.php");
 display_top_tabs($tab_array);
